### PR TITLE
feat: New error handling APIs

### PR DESCRIPTION
**Pull request by hiranya911**
Implements the new error handling APIs proposed at go/firebase-error-handling-go.

* Added the `errorutils` package which exposes the platform error handling functions.
* Added the required shared utilities to `internal/http_client.go` for handling OnePlatform and other HTTP errors.
* Updated the `iid` package to use the new error handling mechanism. I've deprecated the error handling functions currently present in the `iid` package, and pointed them to the new ones introduced in `errorutils`.

go/firebase-error-handling-go

### CODE REVIEW

File: internal/errors.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+		String:    fmt.Sprintf("unexpected http response with status: %d\n%s", resp.Status, string(resp.Body)),
+		Response:  resp.LowLevelResponse(),
+	}
+}
+
+// NewFirebaseErrorOnePlatform parses the response payload as a GCP error response
+// and create an error from the details extracted.
+//
+// If the response failes to parse, or otherwise doesn't provide any useful details
+// CreatePlatformError creates an error with some sensible defaults.
```
Comment (lahirumaramba) [inline internal/errors.go]: It looks like this PR removes `CreatePlatformError()`. Is this comment still valid with those changes?

File: errorutils/errorutils.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+// read-modify-write conflict.
+//
+// This represents an HTTP 409 Conflict status code, without additional information to distinguish
+// between ABORTED or ALREADY_EXISTS error conditions.
+func IsConflict(err error) bool {
+	return internal.HasPlatformErrorCode(err, internal.Conflict)
+}
+
+// IsAborted checks if the given error was due to a concurrency conflict, such as a
+// read-modify-write conflict.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply errorutils/errorutils.go]: I think this is ok. Those 2 error conditions are mutually exclusive in practice. An API may either throw CONFLICT or ABORTED, but never both.

File: errorutils/errorutils.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+// IsUnavailable checks if the given error was caused by an unavailable service.
+//
+// This typically indicates that the target service is temporarily down.
+func IsUnavailable(err error) bool {
+	return internal.HasPlatformErrorCode(err, internal.Unavailable)
+}
+
+// IsDeadlineExceeded checks if the given error was due a request exceeding a deadline.
+//
+// his will happen only if the caller sets a deadline that is shorter than the method's default
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply errorutils/errorutils.go]: Done

File: errorutils/errorutils.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+// IsAborted checks if the given error was due to a concurrency conflict, such as a
+// read-modify-write conflict.
+func IsAborted(err error) bool {
+	return internal.HasPlatformErrorCode(err, internal.Aborted)
+}
+
+// IsAlreadyExists checks if the given error was because a resource that a client tried to create
+// already exists..
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply errorutils/errorutils.go]: Done

File: internal/errors.go (hunk 1 above)
Comment (hiranya911) [reply internal/errors.go]: Updated comment

**Pull request merged**